beacon/beacon-common.js

```diff
@@ -0,0 +1,353 @@
+                    // tests in the session, so fail the entire session.
+                    failSession(session, failure);
+                    return;
+                }
+
+                // The 'stat' call will return an array of zero or more results
+                // of sendbeacon() calls that the server has received and validated.
+                results.forEach(function(result) {
+                    var testCase = session.testCaseLookup[result.id];
+
+                    // While stash.take on the server is supposed to honor read-once, since we're
+                    // polling so frequently it is possible that we will receive the same test result
+                    // more than once.
+                    if (!testCase.done) {
+                        testCase.done = true;
+                        session.doneCount++;
+                    }
+
+                    // Validate that the sendbeacon() was actually sent to the server.
+                    var test = testCase.test;
+                    test.step(function() {
+                        // null JSON values parse as null, not undefined
+                        assert_equals(result.error, null, "'sendbeacon' data must not fail validation");
+                    });
+
+                    test.done();
+                });
+
+                // Continue polling until all of the results come in.
+                if (session.doneCount < session.sentCount) {
+                    // testharness.js frowns upon the use of explicit timeouts, but there is no way
+                    // around the need to poll for these tests, and there is no use spamming the server
+                    // with requestAnimationFrame() just to avoid the use of step_timeout.
+                    step_timeout(waitForResults.bind(this, session), 100);
+                }
+            }).catch(function(error) {
+                failSession(session, `unexpected error reading response, error='${error}'`);
+            });
+        }
+    );
+}
+
+// Fails all of the tests in the session, meant to be called when an infrastructural
+// issue prevents us from deterministically completing the individual tests.
+function failSession(session, reason) {
+    session.testCases.forEach(function(testCase) {
+        var test = testCase.test;
+        test.unreached_func(reason)();
+    });
+}
+
+// Creates an iframe on the document's body and runs the sample tests from the iframe.
+// The iframe is navigated immediately after it sends the data, and the window verifies
+// that the data is still successfully sent.
+//    funcName: "beacon" to send the data via navigator.sendBeacon(),
+//              "fetch" to send the data via fetch() with the keepalive flag.
+function runSendInIframeAndNavigateTests(funcName) {
+    var iframe = document.createElement("iframe");
+    iframe.id = "iframe";
+    iframe.onload = function() {
+        var tests = Array();
+
```

Comment:
In WebKit / Blink, I need to do:
iframe.onload = null;

here, otherwise, the onload handler gets called again for the navigation to Microsoft.com at the end of the test and we run 14 tests instead of 7.